### PR TITLE
Long QuestionAnsweringPipeline fix.

### LINKED ISSUE #16769: Invalid CLS masking in question answer pipelines top K calculation
## Environment info
- `transformers` version: 4.18.0
- Platform: Linux-5.10.102-99.473.amzn2.x86_64-x86_64-with-glibc2.10
- Python version: 3.8.12
- Huggingface_hub version: 0.5.1
- PyTorch version (GPU?): 1.10.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

### Who can help
@Narsil

## Information

Model I am using (Bert, XLNet ...):

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [X] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [X] my own task or dataset: (give details below)

## To reproduce

1. Using any question and answering model which leverages CLS token and run it with pipelines. 
2. Supply an input record which has (con)text longer than the max_seq_length set to cause it to chunk the input record.
3. If you inspect the results of p_mask after line https://github.com/huggingface/transformers/blob/4975002df50c472cbb6f8ac3580e475f570606ab/src/transformers/pipelines/question_answering.py#L307
4. You will see the token position used by CLS in the model (typically the first) has not been set to False and remains True meaning it will be masked out as an undesired token later on.

This will work if the input record is not chunked. The reason is the code before that constructs p_mask doesn't work as expected when the chunk record are different lengths (when calling np.asarray) and the lines pointed to above then fail silently to correctly mask the CLS token position. 

This then causes differences in the calculations of the results answer spans and their associated probabilities.

## Expected behavior

The CLS token position should be correctly set to False so that it is considered a valid token for consideration in answer calculations regardless of whether the input record was chunked or not.

As a simple but possibly not an efficient fix I replaced the two lines in the if pointed to above with the following:

```
    for span_id in range(num_spans):
        cls_index = np.nonzero(np.array(encoded_inputs["input_ids"][span_id]) == self.tokenizer.cls_token_id)
        p_mask[span_id][cls_index] = 0
```

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

This handles CLS index even on splitted QA context.

TODO update the tests for one that actually showcases the bug.

Fixes #16769

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Narsil** [comment]
@LysandreJik  This is not fully ready yet as the test does not really check that the bug is fixed. Do you have any ideas on how to check that behavior ? (Previous code was just silently not doing anything since the np.arrays are an array of `np.list(..)` in the splitted case.)

**Review by LysandreJik — APPROVED** [review]
This looks good to me! If you think the fast test is unnecessary, ok for me to remove it and keep the slow test only.

Feel free to merge when ready!

**Comment by Narsil** [comment]
I'll remove the fast test, don't want to leave a false sense of security here.
